source/cAccountClass.cpp

```diff
@@ -498,7 +498,7 @@ UI16 cAccountClass::CreateAccountSystem( void )
 			}
 			else
 			{
-				if( !r.empty() && r.length() != 0 && oldstrutil::value<SI32>( r ) != INVALIDSERIAL )
+				if( !r.empty() && r.length() != 0 && static_cast<unsigned>(oldstrutil::value<SI32>( r )) != INVALIDSERIAL )
```

Comment:
Can you elaborate more on this, @strattond?

Reply:
Sure.  INVALIDSERIAL is unsigned, which is why you're trying to marry the two up - makes a lot of sense.

I think, rather than cast the oldstrutil call to an unsigned, you can just call it that way direct. e.g.
```suggestion
				if( !r.empty() && r.length() != 0 && oldstrutil::value<UI32>( r ) != INVALIDSERIAL )
```

Because the value() routine looks templated, and this way you'll get back an unsigned int you don't have to recast.

I'll admit, though, I don't know what it'll do if it's expecting the data to come in as -1 from the account system.  So may need a test in there.

Reply:
How does the following look to you? This will also handle when `r` is negative (which will never match `INVALIDSERIAL`).

```cpp
if( !r.empty() && r.length() != 0 && ( oldstrutil::value<SI32>( r ) < 0 || oldstrutil::value<UI32>( r ) != INVALIDSERIAL ))
```

Reply:
This doesn't feel great, as now you're parsing it twice.  I wonder if it's worth getting @punt1959 opinion as he was the one responsible for the template value function I suspect?

Reply:
`r` will only be parsed twice if it does not short circuit during evaluation (worst-case scenario). This also covers the negative case you posed. Furthermore, this would not significantly impact performance if that is your concern. If the value can never be negative, then we can ignore this secondary issue altogether. Hopefully, @punt1959 can confirm whether or not `r` can be negative.

Reply:
The following was the workaround proposed by @punt1959 in Discord. Does this seem reasonable to you, @strattond?
```cpp
r = oldstrutil::trim(r);
if( !r.empty() )
{
  auto temp = oldstrutil::value<std::int64_t>(r);
  if( temp > 0 && temp <= static_cast<std::int64_t>(INVALIDSERIAL) )
  {
    actb.dwCharacters[charNum - 1] = static_cast<UI32>(std::stoul( r, nullptr, 0 ));
    actb.lpCharacters[charNum - 1] = CalcCharObjFromSer( actb.dwCharacters[charNum-1] );
  }
  else
  {
    actb.dwCharacters[charNum - 1] = INVALIDSERIAL;
    actb.lpCharacters[charNum - 1] = nullptr;
  }
}
```

Reply:
@ldilley I'd suspect it should be `temp > 0 && temp < stati...` rather than `<=` as otherwise you'll never fall into the second path?

I'd also add a comment to the cast to int64_t, as it relies on the fact it's keeping it as a large number because of the promotion to a larger type (32 in 64).  If it was a like for like cast, it would fail.

Reply:
Does the edited code in the original post now meet the criteria for both of your points, @strattond?

Reply:
Fixed in 6411368 and per Discord conversation.